x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -187,6 +196,31 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
+    private void logMlNodeHeterogeneity() {
+        ActionListener<Set<String>> architecturesListener = new ActionListener<Set<String>>() {
+            @Override
+            public void onResponse(Set<String> architectures) {
+                if (architectures.size() > 1) {
+                    String architecturesList = String.join(", ", architectures.stream().toList());
```

Comment:
I think any iterable collection works here, so converting to a list is wasteful.

```suggestion
                    String architecturesList = String.join(", ", architectures);
```